core/src/main/java/com/airbnb/aerosolve/core/models/LowRankLinearModel.java

```diff
@@ -0,0 +1,223 @@
+// W: D-by-Y matrix, mapping from label space to the joint embedding
+// Reference: Jason Weston et al. "WSABIE: Scaling Up To Large Vocabulary Image Annotation", IJCAI 2011.
+public class LowRankLinearModel extends AbstractModel {
+
+  static final long serialVersionUID = -8894096678183767660L;
+
+  // featureWeightVector represents the projection from feature space to embedding
+  // Map feature family name, feature name to a column in V
+  // each FloatVector in the map is a D-dim vector
+  @Getter
+  @Setter
+  private Map<String, Map<String, FloatVector>> featureWeightVector;
+
+  // labelWeightVector represents the projection from label space to embedding
+  // Map label to a row in W, each FloatVector in the map is a D-dim vector
+  @Getter
+  @Setter
+  private Map<String, FloatVector> labelWeightVector;
+
+  @Getter
+  @Setter
+  private ArrayList<LabelDictionaryEntry> labelDictionary;
+
+  @Getter
+  @Setter
+  private Map<String, Integer> labelToIndex;
+
+  // size of the embedding
+  @Getter
+  @Setter
+  private int D;
+
+  public LowRankLinearModel() {
+    featureWeightVector = new HashMap<>();
+    labelWeightVector = new HashMap<>();
+    labelDictionary = new ArrayList<>();
+  }
+
+  // In the binary case this is just the score for class 0.
+  // Ideally use a binary model for binary classification.
+  @Override
+  public float scoreItem(FeatureVector combinedItem) {
+    // Not supported.
+    assert (false);
+    Map<String, Map<String, Double>> flatFeatures = Util.flattenFeature(combinedItem);
+    FloatVector sum = scoreFlatFeature(flatFeatures);
+    return sum.values[0];
+  }
+
+  @Override
+  public float debugScoreItem(FeatureVector combinedItem,
+                              StringBuilder builder) {
+    // TODO(peng) : implement debug.
+    return scoreItem(combinedItem);
+  }
+
+
+  @Override
+  public List<DebugScoreRecord> debugScoreComponents(FeatureVector combinedItem) {
+    // TODO(peng): implement debugScoreComponents
+    List<DebugScoreRecord> scoreRecordsList = new ArrayList<>();
+    return scoreRecordsList;
+  }
+
+  public ArrayList<MulticlassScoringResult> scoreItemMulticlass(FeatureVector combinedItem) {
+    ArrayList<MulticlassScoringResult> results = new ArrayList<>();
+    Map<String, Map<String, Double>> flatFeatures = Util.flattenFeature(combinedItem);
+    FloatVector sum = scoreFlatFeature(flatFeatures);
+
+    for (int i = 0; i < labelDictionary.size(); i++) {
+      MulticlassScoringResult result = new MulticlassScoringResult();
+      result.setLabel(labelDictionary.get(i).getLabel());
+      result.setScore(sum.values[i]);
+      results.add(result);
+    }
+
+    return results;
+  }
+
+  public FloatVector scoreFlatFeature(Map<String, Map<String, Double>> flatFeatures) {
+    int dim = labelDictionary.size();
+    FloatVector sum = new FloatVector(dim);
+    FloatVector fvProjection = new FloatVector(D);
+
+    // compute the projection from feature space to D-dim joint space
```

Comment:
I would break these components out into different function calls because they are useful in the future.
e.g. projectFeature.